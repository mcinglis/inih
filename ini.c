
// parts copyright (c) 2009 Brush Technology
//                          <http://brush.co.nz/>
// parts copyright (c) 2015 Positronic Manufacturing
//                          <http://positronicsolar.com.au/>


#include <stdlib.h>
#include <stdio.h>
#include <ctype.h>
#include <string.h>

#include <libtypes/types.h>
#include <libmacro/require.h>

#include "ini.h"


// Strip whitespace chars off end of given string, in place. Return s.
static
char *
rstrip( char * const s )
{
    REQUIRE( s != NULL );

    char * p = s + strlen( s );
    while ( p > s && isspace( ( unsigned char )( *--p ) ) ) {
        *p = '\0';
    }
    return s;
}


// Return a pointer to the first non-whitespace character in `s`.
static
char *
lskip( char * const s )
{
    REQUIRE( s != NULL );

    size_t i = 0;
    while ( isspace( s[ i ] ) ) {
        i++;
    }
    return s + i;
}


// Version of strncpy that ensures dest (size bytes) is null-terminated.
static
char *
strncpy0( char * const dest,
          char const * const src,
          size_t const size )
{
    REQUIRE( dest != NULL, src != NULL );

    strncpy( dest, src, size );
    dest[ size - 1 ] = '\0';
    return dest;
}


static
void
strip_comment( char * const line )
{
    REQUIRE( line != NULL );

    for ( size_t i = 1; line[ i ] != '\0'; i++ ) {
        if ( line[ i ] == ';' && isspace( line[ i - 1 ] ) ) {
            line[ i ] = '\0';
            break;
        }
    }
}


IniError
ini_parse_file( FILE * const file,
                bool ( * const callback )( void * data,
                                           char const * section,
                                           char const * key,
                                           char const * value ),
                void * const data )
{
    REQUIRE( file != NULL, callback != NULL );

    char line_buf[ INI_MAX_LINE_LEN + 1 ]       = { 0 };
    char section[ INI_MAX_SECTION_LEN + 1 ] = { 0 };

    // Step through file line by line:
    int lineno = 0;
    while ( fgets( line_buf, sizeof line_buf, file ) != NULL ) {
        lineno++;

        char * line = line_buf;

        // Check for the leading BOM if any:
        if ( lineno == 1 && ( ( uchar ) line[ 0 ] ) == 0xEF &&
                            ( ( uchar ) line[ 1 ] ) == 0xBB &&
                            ( ( uchar ) line[ 2 ] ) == 0xBF ) {
            line += 3;
        }

        // Strip leading and trailing whitespace:
        line = lskip( rstrip( line ) );

        if ( line[ 0 ] == ';' || line[ 0 ] == '\0' ) {
            continue;
        } else if ( line[ 0 ] == '[' ) {
            // We're parsing a section header line:
            char * const end = strchr( line + 1, ']' );
            if ( end == NULL ) {
                return ( IniError ){ .desc = "unclosed section header",
                                     .line = lineno };
            }
            *end = '\0';
            strncpy0( section, lskip( rstrip( line + 1 ) ), sizeof section );
        } else {
            // Must be a name=value pair:
            strip_comment( line );
            // Separate the key and value:
            char * const eq = strchr( line, '=' );
            if ( eq == NULL ) {
                return ( IniError ){ .desc = "missing assignment characater",
                                     .line = lineno };
            }
            char * const key = line;
            char * const value = lskip( eq + 1 );
            // Terminate `key` at the assignment character:
            *eq = '\0';
            // Strip any trailing whitespace characters:
            rstrip( key );
            rstrip( value );
            // Finally call the callback with this key and value:
            if ( callback( data, section, key, value ) == false ) {
                return ( IniError ){ .desc = "callback failed",
                                     .line = lineno };
            }
        }
    }

    return ferror( file ) ? ( IniError ){ "error reading file" }
                          : ( IniError ){ 0 };
}


IniError
ini_parse( char const * const filename,
           bool ( * const callback )( void * data,
                                      char const * section,
                                      char const * key,
                                      char const * value ),
           void * const data )
{
    REQUIRE( filename != NULL, callback != NULL );

    FILE * const file = fopen(filename, "r");
    if ( file == NULL ) {
        return ( IniError ){ "couldn't open file" };
    }
    IniError const parse_err = ini_parse_file( file, callback, data );
    int const close_err = fclose( file );
    return ( parse_err.desc != NULL ) ? parse_err
         : ( close_err != 0 )         ? ( IniError ){ "couldn't close file" }
                                      : ( IniError ){ 0 };
}

