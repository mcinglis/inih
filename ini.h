
// parts copyright (c) 2009 Brush Technology
//                          <http://brush.co.nz/>
// parts copyright (c) 2015 Positronic Manufacturing
//                          <http://positronicsolar.com.au/>


#ifndef INIH_INI_H
#define INIH_INI_H


#include <stdio.h>
#include <libtypes/types.h>


// Maximum length for lines in an .ini file:
#ifndef INI_MAX_LINE_LEN
#define INI_MAX_LINE_LEN 1024
#endif

// Maximum length of section headers:
#ifndef INI_MAX_SECTION_LEN
#define INI_MAX_SECTION_LEN 64
#endif

// Maximum length of property keys:
#ifndef INI_MAX_KEY_LEN
#define INI_MAX_KEY_LEN 64
#endif


typedef struct ini_error {
    char const * desc;
    size_t line;
} IniError;


// Parse given INI-style file. May have [section]s, name=value pairs
// (whitespace stripped), and comments starting with ';' (semicolon). Section
// is "" if name=value pair parsed before any section heading. name:value
// pairs are also supported as a concession to Python's ConfigParser.
//
// For each name=value pair parsed, call handler function with given user
// pointer as well as section, name, and value (data only valid for duration
// of handler call). Handler should return nonzero on success, zero on error.
//
// Returns 0 on success, line number of first error on parse error (doesn't
// stop on first error), -1 on file open error, or -2 on memory allocation
// error (only when INI_USE_STACK is zero).
IniError
ini_parse( char const * filename,
           bool ( * callback )( void * data,
                                char const * section,
                                char const * key,
                                char const * value ),
           void * data );


// Same as `ini_parse()`, but takes a `FILE *` instead of filename. This
// doesn't close the file when it's finished -- the caller must do that.
IniError
ini_parse_file( FILE * file,
                bool ( * callback )( void * data,
                                     char const * section,
                                     char const * key,
                                     char const * value ),
                void * data );


#endif

