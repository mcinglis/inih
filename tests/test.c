
// Run these tests by doing something like:
//
//     $ diff stdout_expected.txt <(./unittest)
//
// The tests pass if there was no diff output.

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "../ini.h"


int count = 1;


bool dumper( void * const vn,
             char const * const section,
             char const * const key,
             char const * const value)
{
    int * const n = vn;
    *n = count;
    count++;

    printf( "[%s] \"%s\" = \"%s\"\n", section, key, value );
    return true;
}


void
parse( char const * const filepath ) {
    int n = 0;
    printf( "%s:\n", filepath );
    IniError const err = ini_parse( filepath, dumper, &n );
    printf( "%s: n = %d\n", filepath, n );
    if ( err.desc != NULL ) {
        printf( "%s: error at line %zu: %s\n",
                filepath, err.line, err.desc );
    }
    printf( "------\n" );
}


int
main( void )
{
    parse( "no_file.ini" );
    parse( "normal.ini" );
    parse( "bad_section.ini" );
    parse( "bad_comment.ini" );
    parse( "user_error.ini" );
    parse( "bom.ini" );
    parse( "colon.ini" );
}
