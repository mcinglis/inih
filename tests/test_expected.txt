no_file.ini:
no_file.ini: n = 0
no_file.ini: error at line 0: couldn't open file
------
normal.ini:
[section1] "one" = "This is a test"
[section1] "two" = "1234"
[section 2] "happy" = "4"
[section 2] "sad" = ""
[comment_test] "test1" = "1;2;3"
[comment_test] "test2" = "2;3;4;this won't be a comment, needs whitespace before ';'"
[comment_test] "test;3" = "345"
[comment_test] "test4" = "4#5#6"
[comment_test] "#test5" = "567"
[comment_test] "# test6" = "678"
normal.ini: n = 10
------
bad_section.ini:
[section1] "name1" = "value1"
bad_section.ini: n = 11
bad_section.ini: error at line 3: unclosed section header
------
bad_comment.ini:
bad_comment.ini: n = 0
bad_comment.ini: error at line 1: missing assignment characater
------
user_error.ini:
[section] "a" = "b"
[section] "user" = "parse_error"
[section] "c" = "d"
user_error.ini: n = 14
------
bom.ini:
[bom_section] "bom_name" = "bom_value"
[bom_section] "key“" = "value“"
bom.ini: n = 16
------
colon.ini:
colon.ini: n = 0
colon.ini: error at line 2: missing assignment characater
------
